### PR TITLE
WordStack optimizations

### PR DESCRIPTION
**Pull request by PetarMax**
This PR introduces some optimizations that should allows us to manipulate word stacks more efficiently..

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md [hunk 1]
```diff
@@ -238,71 +238,185 @@ A cons-list is used for the EVM wordstack.
...
-    rule [#take.zero-pad]:  #take(N, .WordStack)         => 0 : #take(N -Int 1, .WordStack) requires N >Int 0
-    rule [#take.recursive]: #take(N, (W : WS):WordStack) => W : #take(N -Int 1, WS)         requires N >Int 0
+ // Expected use cases are from 0-4
+    rule [ws-take.0]: #take(0,                      _) => .WordStack
+    rule [ws-take.1]: #take(1, W0 :                 _) => W0 : .WordStack
+    rule [ws-take.2]: #take(2, W0 : W1 :            _) => W0 : W1 : .WordStack
+    rule [ws-take.3]: #take(3, W0 : W1 : W2 :       _) => W0 : W1 : W2 : .WordStack
+    rule [ws-take.4]: #take(4, W0 : W1 : W2 : W3 :  _) => W0 : W1 : W2 : W3 : .WordStack
+ // For unexpected cases
+    rule [ws-take.N]: #take(N,  _ :  _ :  _ :  _ : WS) => #take(N -Int 4, WS) requires 4 <Int N
```
Comment (gtrepta) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md:254]:
```suggestion
    rule [ws-take.N]: #take(N,  W0 :  W1 :  W2 :  W3 : WS) => W0 : W1 : W2 : W3 : #take(N -Int 4, WS) requires 4 <Int N
```

I think it should be this?

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md [hunk 2]
```diff
@@ -238,71 +238,185 @@ A cons-list is used for the EVM wordstack.
...
-    rule #drop(N, (W : WS):WordStack) => #drop(1, #drop(N -Int 1, (W : WS))) requires         N >Int 1
-    rule #drop(1, (_ : WS):WordStack) => WS
+ // Expected use cases are from 0-4
+    rule [ws-drop.0]: #drop(0,                     WS) => WS
+    rule [ws-drop.1]: #drop(1, _ :                 WS) => WS
+    rule [ws-drop.2]: #drop(2, _ : _ :             WS) => WS
+    rule [ws-drop.3]: #drop(3, _ : _ : _ :         WS) => WS
+    rule [ws-drop.4]: #drop(4, _ : _ : _ : _ :     WS) => WS
+ // For unexpected cases
+    rule [ws-drop.N]: #drop(N, _ : _ : _ : _ : _ : WS) => #drop(N -Int 4, WS) requires 4 <Int N
```
Comment (gtrepta) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md:267]:
```suggestion
    rule [ws-drop.N]: #drop(N, _ : _ : _ : _ : WS) => #drop(N -Int 4, WS) requires 4 <Int N
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md [hunk 3]
```diff
@@ -238,71 +238,185 @@ A cons-list is used for the EVM wordstack.
...
+    rule [ws-size-09]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 9
+    rule [ws-size-10]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 10
+    rule [ws-size-11]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 11
+    rule [ws-size-12]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 12
+    rule [ws-size-13]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 13
+    rule [ws-size-14]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 14
+    rule [ws-size-15]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 15
+    rule [ws-size-16]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 16
+
+    rule #sizeWordStack ( _ : WS ) => 1 +Int #sizeWordStack(WS)
```
Comment (gtrepta) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md:419]:
```suggestion
    rule [ws-size-N]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : WS) => 16 +Int #sizeWordStack(WS) requires notBool WS ==K .WordStack
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k [hunk 4]
```diff
@@ -46,7 +46,9 @@ module LEMMAS-WITHOUT-SLOT-UPDATES [symbolic]
   // Word Reasoning
   // ########################
 
-    rule 0 <=Int #sizeWordStack ( _ , N ) => true requires 0 <=Int N [simplification, smt-lemma]
+    rule #sizeWordStack(WS [ I := _ ]) => #sizeWordStack(WS) requires I <Int #sizeWordStack(WS) [simplification]
```
Comment (gtrepta) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k:49]: Would it be better to have `concrete(I)` on this?

File: optimizer/optimizations.md [hunk 5]
```diff
@@ -10,10 +10,9 @@ requires "lemmas/int-simplification.k"
 module EVM-OPTIMIZATIONS-LEMMAS [symbolic]
     imports EVM
 
-    rule #sizeWordStack(WS           , N) => #sizeWordStack(WS, 0) +Int N requires N =/=Int 0                [simplification]
-    rule #sizeWordStack(WS [ I := _ ], N) => #sizeWordStack(WS, N)        requires I <Int #sizeWordStack(WS) [simplification]
-    rule 0 <=Int #sizeWordStack(_ , 0)    => true                                                            [simplification]
-    rule #sizeWordStack(_ , 0) <Int N     => false                        requires N <=Int 0                 [simplification]
+    rule #sizeWordStack(WS [ I := _ ]) => #sizeWordStack(WS, N) requires I <Int #sizeWordStack(WS) [simplification]
```
Comment (gtrepta) [inline optimizer/optimizations.md:13]: What is `N` here?

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md (hunk 1 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md:254]: Yes.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md (hunk 2 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md:267]: Yes.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k (hunk 4 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k:49]: No, it needs to be able to survive symbolic ones as well.

File: optimizer/optimizations.md (hunk 5 above)
Comment (PetarMax) [reply optimizer/optimizations.md:13]: Leftover, removed.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md (hunk 3 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md:419]: We need `#sizeWordStack ( _ : WS ) => 1 +Int #sizeWordStack(WS)` for when `WS` ends in a symbolic word stack, not all end in `.WordStack`.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md [hunk 6]
```diff
@@ -1171,6 +1171,7 @@ These operators query about the current return data buffer.
          <localMem> LM </localMem>
          <log> L => L ListItem({ ACCT | WordStack2List(#take(N, WS)) | #range(LM, MEMSTART, MEMWIDTH) }) </log>
       requires #sizeWordStack(WS) >=Int N
+      [preserves-definedness]
 ```
```
Comment (ehildenb) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md:1175]: Why does this rule fail the `preserves-definedness` check in the first place?

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md [hunk 7]
```diff
@@ -238,71 +238,184 @@ A cons-list is used for the EVM wordstack.
...
+    rule [ws-size-09]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 9
+    rule [ws-size-10]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 10
+    rule [ws-size-11]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 11
+    rule [ws-size-12]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 12
+    rule [ws-size-13]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 13
+    rule [ws-size-14]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 14
+    rule [ws-size-15]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 15
+    rule [ws-size-16]: #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : .WordStack) => 16
+    rule [ws-size-N]:  #sizeWordStack (_ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : _ : WS) => 16 +Int #sizeWordStack(WS)
+    rule [ws-size-O]:  #sizeWordStack (_ : WS) => 1 +Int #sizeWordStack(WS) [priority(75)]
```
Comment (ehildenb) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md:418]: I think the reason for how it was structured before was to keep the computation tail-recursive, which the LLVM backend requires for efficient memory/time usage. Can we check the LLVM backend performance on teh conformance test-suite here?

File: optimizer/optimizations.md [hunk 8]
```diff
@@ -23,8 +22,40 @@ module EVM-OPTIMIZATIONS
...
+          SCHED
+        </schedule>
+        <useGas>
+          USEGAS
+        </useGas>
+        <ethereum>
+          <evm>
+            <callState>
+              <wordStack>
+                ( W0 : W1 : WS => chop( ( W0 +Int W1 ) ) : WS )
```
Comment (ehildenb) [inline optimizer/optimizations.md:41]: Why is this added here?

We already have a test-harness that takes the rules in this file and discharges them as claims directly.

Comment (ehildenb) [reply optimizer/optimizations.md:41]: Oh, actually, this entire directory isn't used at all anymore. We really should delete it, but I left it for illustrative purposes. Any reason to add this?

File: tests/failing-symbolic.haskell-booster-dev [hunk 9]
```diff
@@ -135,4 +135,8 @@ tests/specs/mcd-structured/flopper-tick-pass-rough-spec.k
 tests/specs/mcd-structured/functional-spec.k
 tests/specs/mcd-structured/pot-join-pass-rough-spec.k
 tests/specs/mcd-structured/vow-fess-fail-rough-spec.k
-tests/specs/mcd-structured/vow-flog-fail-rough-spec.k
\ No newline at end of file
+tests/specs/mcd-structured/vow-flog-fail-rough-spec.k
+tests/specs/mcd/dai-adduu-fail-rough-spec.k
+tests/specs/mcd/flipper-addu48u48-fail-rough-spec.k
+tests/specs/mcd-structured/dai-adduu-fail-rough-spec.k
+tests/specs/mcd-structured/flipper-addu48u48-fail-rough-spec.k
```
Comment (ehildenb) [inline tests/failing-symbolic.haskell-booster-dev:142]: Well that's less than ideal I guess.....

**Comment by ehildenb** [comment]
We really need some performance metrics posted here, both for Haskell and LLVM backend.

File: tests/failing-symbolic.haskell-booster-dev (hunk 9 above)
Comment (PetarMax) [reply tests/failing-symbolic.haskell-booster-dev:142]: It is, but the booster will adapt and be more expressive, as per [this issue](https://github.com/runtimeverification/haskell-backend/issues/4053).

Comment (PetarMax) [reply tests/failing-symbolic.haskell-booster-dev:142]: This might allow us to prove further tests using just `booster-dev`.

File: optimizer/optimizations.md (hunk 8 above)
Comment (PetarMax) [reply optimizer/optimizations.md]: Removed.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md (hunk 6 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md]: Removed.

**Comment by PetarMax** [comment]
What would be the best way to get the metrics? For the LLVM backend, I guess the conformance test suite.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md (hunk 7 above)
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md]: I don't like this, I've reverted back to just the simple tail recursion.

@ehildenb Do you remember why there was an auxiliary function here with two parameters instead of a direct computation?

Comment (gtrepta) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md]: https://github.com/runtimeverification/evm-semantics/blob/3d7c9c2b9419a663af0f7e3e8dfd4da50ee16892/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md?plain=1#L398-L401

This isn't tail recursive. It's doing an add operation after the call. That's why there was an auxiliary function with the extra parameter.

Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md]: Ah yes yes, sorry.

Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md]: I will bring it back. It's not the thing that's breaking the proofs anyways and works well.

If we wanted to speed this calculation up, we can merge something like [this](https://github.com/runtimeverification/evm-semantics/pull/2628).

**Comment by geo2a** [comment]
@PetarMax regarding the performance metrics for the Haskell backend: please ask @anvacaru, I remember that he has adapted our Nix-based performance measurement script to vary the version of KEVM instead of HB. This thould get tou a table comparing KEVM `master` vs `branch`, while keeping the K and HB versions fixed.

**Comment by PetarMax** [comment]
Just to say regarding concrete execution and the LLVM backend - the moment that we have anything symbolic in the word stack, we cannot use the LLVM backend to compute anything over it, even if it doesn't depend on the symbolicness (say, length), and so any tail recursion optiimisations will only work for purely concrete word stacks.

**Comment by anvacaru** [comment]
Yes, I put something together from the [Haskell Backend's test-performance-kevm.sh](https://github.com/runtimeverification/haskell-backend/blob/master/scripts/performance-tests-kevm.sh) script. 

I would run this [profile-kevm](https://github.com/anvacaru/git-helpers/blob/master/scripts/profile-kevm) script for both branches I'm interested in.
```sh
export KEVM_VERSION=master; profile-kevm
export KEVM_VERSION=wordstack-optimizations; profile-kevm
```
then I would use the [compare_logs.py logfile_1 logfile_2 ](https://github.com/anvacaru/git-helpers/blob/master/scripts/python/bin/compare_logs.py)to generate the table with the time values. 

I will give it a go in the background and post the results here.

**Comment by anvacaru** [comment]
It seems like these optimizations are causing a slowdown. @PetarMax can you confirm this?


| Test                                                                                      | master-1bd8e09de time | wordstack-optimizations-3801e920c time | (master-1bd8e09de/wordstack-optimizations-3801e920c) time 
|-------------------------------------------------------------------------------------------|-----------------------|----------------------------------------|-----------------------------------------------------------
| examples/sum-to-n-spec.k                                                                  | 91.79                 | 103.43                                 | 0.8874601179541719                                        
| mcd/end-subuu-pass-spec.k                                                                 | 55.36                 | 60.03                                  | 0.9222055638847243                                        
| mcd-structured/end-subuu-pass-spec.k                                                      | 55.54                 | 59.61                                  | 0.9317228652910585                                        
| erc20/ds/transferFrom-failure-1-b-spec.k                                                  | 122.0                 | 130.18                                 | 0.937163926870487                                         
| mcd-structured/dai-adduu-fail-rough-spec.k                                                | 48.6                  | 51.55                                  | 0.9427740058195927                                        
| kontrol/test-expectreverttest-testfail_expectrevert_false-0-spec.k                        | 63.9                  | 67.69                                  | 0.9440094548677795                                        
| erc20/ds/transferFrom-failure-1-c-spec.k                                                  | 99.87                 | 105.54                                 | 0.9462762933484935                                        
| mcd/dai-adduu-fail-rough-spec.k                                                           | 48.6                  | 51.33                                  | 0.9468147282291058                                        
| benchmarks/functional-spec.k                                                              | 357.99                | 376.78                                 | 0.9501300493656777                                        
| erc20/ds/transferFrom-failure-2-c-spec.k                                                  | 47.24                 | 49.7                                   | 0.9505030181086519                                        
| mcd-structured/flipper-addu48u48-fail-rough-spec.k                                        | 48.03                 | 50.49                                  | 0.9512774806892453                                        
| kontrol/test-emitcontracttest-testexpectemitcheckemitter-0-spec.k                         | 47.64                 | 50.05                                  | 0.951848151848152                                         
| mcd/flipper-addu48u48-fail-rough-spec.k                                                   | 47.99                 | 50.36                                  | 0.9529388403494837                                        
| erc20/ds/balanceOf-spec.k                                                                 | 46.33                 | 48.55                                  | 0.9542739443872297                                        
| mcd-structured/functional-spec.k                                                          | 324.39                | 339.38                                 | 0.9558312216394602                                        
| benchmarks/storagevar00-spec.k                                                            | 42.9                  | 44.76                                  | 0.9584450402144772                                        
| kontrol/test-arithmetictest-test_wmul_weakly_increasing_positive-uint256-uint256-0-spec.k | 67.85                 | 70.71                                  | 0.9595531042285391                                        
| benchmarks/overflow00-overflow-spec.k                                                     | 43.72                 | 45.55                                  | 0.9598243688254665                                        
| mcd-structured/cat-exhaustiveness-spec.k                                                  | 104.35                | 108.7                                  | 0.9599816007359705                                        
| benchmarks/ecrecover00-sigvalid-spec.k                                                    | 53.29                 | 55.48                                  | 0.9605263157894737                                        
| mcd-structured/flopper-tick-pass-rough-spec.k                                             | 83.28                 | 86.64                                  | 0.961218836565097                                         
| kontrol/test-countertest-testincrement-0-spec.k                                           | 85.3                  | 88.7                                   | 0.9616685456595264                                        
| examples/solidity-code-spec.md                                                            | 105.59                | 109.78                                 | 0.9618327564219348                                        
| kontrol/test-expectreverttest-test_expectrevert_bytes4-0-spec.k                           | 63.52                 | 66.04                                  | 0.9618413082980012                                        
| kontrol/test-expectreverttest-testfail_expectrevert_failandsuccess-0-spec.k               | 46.77                 | 48.54                                  | 0.9635352286773795                                        
| benchmarks/requires01-a0le0-spec.k                                                        | 43.32                 | 44.95                                  | 0.9637374860956618                                        
| mcd/functional-spec.k                                                                     | 548.15                | 568.02                                 | 0.9650188373648815                                        
| mcd-structured/flopper-file-pass-rough-spec.k                                             | 17.9                  | 17.28                                  | 1.0358796296296295                                        
| mcd/flopper-dent-guy-same-pass-rough-spec.k                                               | 641.56                | 610.61                                 | 1.0506870179001326                                        
| TOTAL                                                                                     | 3452.77               | 3560.43                                | 0.9697620792994105
